R/compare_df_types.R

```diff
@@ -0,0 +1,245 @@
+#'   \code{class_colname}) defining the classes using
+#'   \code{compare_df_types_class_detect()}.
+#' @noRd
+compare_df_types_df_maker <- function(x, class_colname="class")
+  UseMethod("compare_df_types_df_maker")
+
+compare_df_types_df_maker.data.frame <- function(x, class_colname="class") {
+  if (class_colname == "column_name") {
+    stop('`class_colname` cannot be "column_name"')
+  }
+  if (ncol(x) == 0) {
+    warning(class_colname, " has zero columns and will not appear in output.")
+    ret <- data.frame(column_name=character(0), stringsAsFactors=FALSE)
+  } else {
+    ret <-
+      data.frame(
+        column_name=names(x),
+        X=sapply(X=x, FUN=compare_df_types_class_detect),
+        stringsAsFactors=FALSE
+      )
+    names(ret)[2] <- class_colname
+  }
+  ret
+}
+
+compare_df_types_df_maker.list <- function(x, class_colname="class") {
+  if (length(class_colname) != length(x)) {
+    stop("`x` and `class_colname` must be the same length.")
+  } else if (any(class_colname == "column_name")) {
+    stop('`class_colname` cannot be "column_name"')
+  }
+  ret <-
+    lapply(
+      X=seq_along(x),
+      FUN=function(idx) {
+        compare_df_types_df_maker(x=x[[idx]], class_colname=class_colname[[idx]])
+      }
+    )
+  Reduce(f=function(x, y) {merge(x, y, by="column_name", all=TRUE)}, x=ret)
+}
+
+#' Will row binding succeed?
+#' @inheritParams compare_df_types
+#' @param verbose Print the mismatching columns if binding will fail.
+#' @return \code{TRUE} if row binding will succeed or \code{FALSE} if it will
+#'   fail.
+#' @family Data frame type comparison
+#' @examples
+#' compare_df_types_success(data.frame(A=1), data.frame(A=2))
+#' compare_df_types_success(data.frame(A=1), data.frame(B=2))
+#' compare_df_types_success(data.frame(A=1), data.frame(B=2), verbose=FALSE)
+#' compare_df_types_success(data.frame(A=1), data.frame(B=2), bind_check="bind_rows")
+#' @export
+compare_df_types_success <- function(..., return="mismatch", bind_check=c("rbind", "bind_rows"), verbose=TRUE) {
+  return <- match.arg(return)
+  bind_check <- match.arg(bind_check)
+  ret <- compare_df_types(..., return=return, bind_check=bind_check)
+  if (nrow(ret) & verbose) {
+    print(ret)
+  }
+  nrow(ret) == 0
+}
+
+#' Describe the class(es) of an object
+#'
+#' @details An S3 generic method can be written for
+#'   \code{compare_df_types_class_detect()} for other types that may need more
+#'   definition than the default method.
+#'
+#' @param x The object to describe
+#' @return A character scalar describing the class(es) of an object where if the
+#'   scalar will match, columns in a data.frame (or similar object) should bind
+#'   together without issue.
+#' @family Data frame type comparison
+#' @export
+compare_df_types_class_detect <- function(x) {
+  UseMethod("compare_df_types_class_detect")
+}
+
+#' @describeIn compare_df_types_class_detect Describe factors with their levels
```

Comment:
More of a general comment than this line: I notice that `compare_df_types` flags mismatched factor levels and `compare_df_types_success` is `FALSE` in that situation.  But `rbind` with different factor levels succeeds - seems to set the levels to the union of the levels? - and `bind_rows` succeeds though with a warning and coercion to character.

I'm not sure about returning `FALSE` for bindable when in fact they are bindable.  But, I do agree it's worth capturing mismatched levels.  Do you have any ideas for how to have it both ways?  Only thing I can think of is another argument called like `strict_levels` or `strict_factors` that would be `TRUE` or `FALSE`.

Reply:
This is a good question that I'd not considered.  I think it needs a bit more consideration.

The best fit would be to have a general argument passed from `describe_df_cols()` to `describe_class()` because that is where the code impact would be (it would change how `describe_class()` works for factors).  A challenge implementing this change that is that we are already using `...` for the main inputs.  Factors are special in that they are commonly used and part of base R, so maybe the special argument makes sense.

As I've typed the above, I think that I'll add a `strict_description` which will be passed to `describe_class()`.  If `TRUE`, factors will show up with levels and "ordered".  If `FALSE`, factors and ordered factors will show up as just "factor".  This won't fully address all bindable combinations (character/factor can bind), but the user will typically be warned of that otherwise.